README.md
# AutoMoLi - **Auto**matic **Mo**tion **Li**ghts

[![python_badge](https://img.shields.io/static/v1?label=python&message=3.8%20|%203.9&color=blue&style=flat)](https://www.python.org) [![hacs_badge](https://img.shields.io/badge/HACS-Default-orange.svg)](https://github.com/hacs/integration)

Fully *automatic light management* based on motion as [AppDaemon](https://github.com/home-assistant/appdaemon) app.  

## Features

* multiple `daytimes` to define different scenes for morning, noon, ...
* supports **Hue** (for Hue Rooms/Groups) & **Home Assistant** scenes
* switches lights **and** plugs (with lights)
* supports **illumination sensors** to switch the light just if needed
* supports **humidity sensors** as blocker (the "*shower case*")

## Try it! On `amd64`, `arm` and `arm64`!

You can try AutoMoLi via [Docker](https://hub.docker.com/r/benleb/automoli) without installing anything! The Image is the default AppDaemon one with AutoMoLi, and a simple default configuration added. See the [AppDaemon Docker Tutorial](https://appdaemon.readthedocs.io/en/latest/DOCKER_TUTORIAL.html) on how to use it in general.  

AutoMoLi is built around `rooms`. For the auto-discovery of your lights and sensors to work, AutoMoLi expects motion sensors and lights including a `room` name. The exact patterns are listed in [Auto-Discovery of Lights and Sensors](https://github.com/benleb/ad-automoli#auto-discovery-of-lights-and-sensors) You can set a `room` with the **AUTOMOLI_ROOM** variable in the Docker *run* command.

```bash
docker run --rm --interactive --tty --name AutoMoLi \
--env HA_URL="<HA URL>" \
--env TOKEN="<HA Token>" \
--env AUTOMOLI_ROOM="bathroom" \
--ports 5050:5050 \
benleb/automoli:latest
```

Port 5050 is opened to give access to the AppDaemon Admin-UI at <http://127.0.0.1:5050>

### Example

To test AutoMoLi in your **Esszimmer** (german for dining room), use `... --env AUTOMOLI_ROOM="esszimmer" ...` in the Docker *run* command.

* AppDaemon will show you its config file on startup: ![cfg](https://raw.githubusercontent.com/benleb/ad-automoli/master/.github/cfg.png)

* If everything works, AutoMoLi will show you the configuration it has parsed, including the discovered sensors: ![cfg-loaded](https://raw.githubusercontent.com/benleb/ad-automoli/master/.github/cfg_loaded.png)

* This is how it looks when AutoMoLi manages your lights: ![running](https://raw.githubusercontent.com/benleb/ad-automoli/master/.github/run.png)

## Installation

Use [HACS](https://github.com/hacs/integration) or [download](https://github.com/benleb/ad-automoli/releases) the `automoli` directory from inside the `apps` directory here to your local `apps` directory, then add the configuration to enable the `automoli` module.

## Requirements/Usage

### *some things are not yet documented here but the code is commented*

* This must be loaded/configured for every **`room`** separately, see example configuration.

## Auto-Discovery of Lights and Sensors

If sensors/lights entities are in this form:

* ***sensor.illumination_`room`***
* ***binary_sensor.motion_sensor_`room`***
* ***binary_sensor.motion_sensor_`room`_something***
* ***light.`room`***  

AutoMoLi will detect them automatically.
Manually configured entities take precedence.

## App configuration

Add your configuration to appdaemon/apps/apps.yaml, an example is below.

```yaml
livingroom:
  module: automoli
  class: AutoMoLi
  room: livingroom
  disable_switch_entities:
    - input_boolean.automoli
    - input_boolean.disable_my_house
  delay: 600
  daytimes:
    # This rule "morning" uses a scene, the scene.livingroom_morning Home Assistant scene will be used
    - { starttime: "05:30", name: morning, light: "scene.livingroom_morning" }
    - { starttime: "07:30", name: day, light: "scene.livingroom_working" }
    # This rule"evening" uses a percentage brightness value, and the lights specified in lights: below will be set to 90%
    - { starttime: "20:30", name: evening, light: 90 }
    - { starttime: "22:30", name: night, light: 20 }
    # This rule has the lights set to 0, so they will no turn on during this time period
    - { starttime: "23:30", name: more_night, light: 0 }
  # If you are using an illuminance sensor you can set the lowest value here that blocks the lights turning on if its already light enough
  illuminance: sensor.illuminance_livingroom
  illuminance_threshold: 100
  # You can specify a light group or list of lights here
  lights:
    - light.livingroom
  # You can specify a list of motion sensors here
  motion:
    - binary_sensor.motion_sensor_153d000224f421
    - binary_sensor.motion_sensor_128d4101b95fb7
  # See below for info on humidity
  humidity:
    - sensor.humidity_128d4101b95fb7

bathroom:
  module: automoli
  class: AutoMoLi
  room: bathroom
  delay: 180
  motion_state_on: "on"
  motion_state_off: "off"
  daytimes:
    - { starttime: "05:30", name: morning, light: 45 }
    - { starttime: "07:30", name: day, light: "Day" }
    - { starttime: "20:30", name: evening, light: 100 }
    - { starttime: "22:30", name: night, light: 0 }
  # As this is a bathroom there could be the case that when taking a bath or shower, motion is not detected and the lights turn off, which isnt helpful, so the following settings allow you to use a humidity sensor and humidity threshold to prevent this by detecting the humidity from the shower and blocking the lights turning off.
  humidity:
    - sensor.humidity_128d4101b95fb7
  humidity_threshold: 75
  lights:
    - light.bathroom
    - switch.plug_68fe8b4c9fa1
  motion:
    - binary_sensor.motion_sensor_158d033224e141
```

key | optional | type | default | description
-- | -- | -- | -- | --
`module` | False | string | automoli | The module name of the app.
`class` | False | string | AutoMoLi | The name of the Class.
`room` | False | string | | The "room" used to find matching sensors/light
`disable_switch_entities` | True | list/string | | One or more Home Assistant Entities as switch for AutoMoLi. If the state of **any** entity is *off*, AutoMoLi is *deactivated*. (Use an *input_boolean* for example)
`disable_switch_states` | True | list/string | ["off"] | Custom states for `disable_switch_entities`. If the state of **any** entity is *in this list*, AutoMoLi is *deactivated*. Can be used to disable with `media_players` in `playing` state for example.
`disable_hue_groups` | False | boolean | | Disable the use of Hue Groups/Scenes
`delay` | True | integer | 150 | Seconds without motion until lights will switched off. Can be disabled (lights stay always on) with `0`
~~`motion_event`~~ | ~~True~~ | ~~string~~ | | **replaced by `motion_state_on/off`**
`daytimes` | True | list | *see code* | Different daytimes with light settings (see below)
`lights` | True | list/string | *auto detect* | Light entities
`motion` | True | list/string | *auto detect* | Motion sensor entities
`illuminance` | True | list/string |  | Illuminance sensor entities
`illuminance_threshold` | True | integer |  | If illuminance is *above* this value, lights will *not switched on*
`humidity` | True | list/string |  | Humidity sensor entities
`humidity_threshold` | True | integer |  | If humidity is *above* this value, lights will *not switched off*
`motion_state_on` | True | integer | | If using motion sensors which don't send events if already activated, like Xiaomi do, add this to your config with "on". This will listen to state changes instead
`motion_state_off` | True | integer | | If using motion sensors which don't send events if already activated, like Xiaomi do, add this to your config with "off". This will listen to the state changes instead.

### daytimes

key | optional | type | default | description
-- | -- | -- | -- | --
`starttime` | False | string | | Time this daytime starts
`name` | False | string | | A name for this daytime
`delay` | True | integer | 150 | Seconds without motion until lights will switched off. Can be disabled (lights stay always on) with `0`. Setting this will overwrite the global `delay` setting for this daytime.
`light` | False | integer/string | | Light setting (percent integer value (0-100) in or scene entity


CHANGELOG.md
# Change Log

## [v0.8.1](https://github.com/benleb/ad-automoli/tree/v0.8.1) (2020-08-03)
[Full Changelog](https://github.com/benleb/ad-automoli/compare/v0.8.0...v0.8.1)

**Closed issues:**

- Feature Request [\#46](https://github.com/benleb/ad-automoli/issues/46)

## [v0.8.0](https://github.com/benleb/ad-automoli/tree/v0.8.0) (2020-08-03)
[Full Changelog](https://github.com/benleb/ad-automoli/compare/v0.7.3...v0.8.0)

**Closed issues:**

- illuminance\_threshold cause error [\#45](https://github.com/benleb/ad-automoli/issues/45)
- ValueError: could not convert string to float: 'unavailable' [\#44](https://github.com/benleb/ad-automoli/issues/44)
- Support Python 3.7 [\#43](https://github.com/benleb/ad-automoli/issues/43)
- Unable to find module - Example configuration [\#36](https://github.com/benleb/ad-automoli/issues/36)
- how is scene suppoesd to work [\#35](https://github.com/benleb/ad-automoli/issues/35)

## [v0.7.3](https://github.com/benleb/ad-automoli/tree/v0.7.3) (2020-06-29)
[Full Changelog](https://github.com/benleb/ad-automoli/compare/v0.7.2...v0.7.3)

**Closed issues:**

- AttributeError: 'list' object has no attribute 'split' [\#42](https://github.com/benleb/ad-automoli/issues/42)

## [v0.7.2](https://github.com/benleb/ad-automoli/tree/v0.7.2) (2020-05-19)
[Full Changelog](https://github.com/benleb/ad-automoli/compare/v0.7.1...v0.7.2)

## [v0.7.1](https://github.com/benleb/ad-automoli/tree/v0.7.1) (2020-05-19)
[Full Changelog](https://github.com/benleb/ad-automoli/compare/v0.7.0...v0.7.1)

## [v0.7.0](https://github.com/benleb/ad-automoli/tree/v0.7.0) (2020-05-19)
[Full Changelog](https://github.com/benleb/ad-automoli/compare/v0.6.1...v0.7.0)

**Closed issues:**

- Daytime is using UTC [\#41](https://github.com/benleb/ad-automoli/issues/41)
- Gledopto lights. [\#33](https://github.com/benleb/ad-automoli/issues/33)

**Merged pull requests:**

- Patch 2 [\#31](https://github.com/benleb/ad-automoli/pull/31) ([samnewman86](https://github.com/samnewman86))

## [v0.6.1](https://github.com/benleb/ad-automoli/tree/v0.6.1) (2020-02-29)
[Full Changelog](https://github.com/benleb/ad-automoli/compare/v0.6.0...v0.6.1)

**Implemented enhancements:**

- Lights turn off even when disable switch is turned off [\#32](https://github.com/benleb/ad-automoli/issues/32)

**Closed issues:**

- Allow sunset/sunrise + offset for daytimes [\#30](https://github.com/benleb/ad-automoli/issues/30)
- Lights do not turnOff direct when illuminate is more than set value [\#27](https://github.com/benleb/ad-automoli/issues/27)
- App not loading [\#26](https://github.com/benleb/ad-automoli/issues/26)
- Xiaomi presence detectors cannot be used for illuminance calculation [\#19](https://github.com/benleb/ad-automoli/issues/19)

**Merged pull requests:**

- Updated readme [\#25](https://github.com/benleb/ad-automoli/pull/25) ([samnewman86](https://github.com/samnewman86))

## [v0.6.0](https://github.com/benleb/ad-automoli/tree/v0.6.0) (2020-02-14)
[Full Changelog](https://github.com/benleb/ad-automoli/compare/v0.5.5...v0.6.0)

**Fixed bugs:**

- ModuleNotFoundError from ad-ench applies here, too [\#15](https://github.com/benleb/ad-automoli/issues/15)

## [v0.5.5](https://github.com/benleb/ad-automoli/tree/v0.5.5) (2020-02-13)
[Full Changelog](https://github.com/benleb/ad-automoli/compare/v0.5.4...v0.5.5)

## [v0.5.4](https://github.com/benleb/ad-automoli/tree/v0.5.4) (2020-02-13)
[Full Changelog](https://github.com/benleb/ad-automoli/compare/v0.5.3...v0.5.4)

**Implemented enhancements:**

- Feature Idea: daytimes to configure on/off, delay, etc. [\#6](https://github.com/benleb/ad-automoli/issues/6)

## [v0.5.3](https://github.com/benleb/ad-automoli/tree/v0.5.3) (2020-02-09)
[Full Changelog](https://github.com/benleb/ad-automoli/compare/v0.5.2...v0.5.3)

## [v0.5.2](https://github.com/benleb/ad-automoli/tree/v0.5.2) (2020-02-09)
[Full Changelog](https://github.com/benleb/ad-automoli/compare/v0.5.1...v0.5.2)

**Closed issues:**

- \[HELP\] Where should i put config ? [\#18](https://github.com/benleb/ad-automoli/issues/18)
- Light not turned on although automoli tells it did that [\#16](https://github.com/benleb/ad-automoli/issues/16)

apps/automoli/automoli.py
"""AutoMoLi.
   Automatic Motion Lights

  @benleb / https://github.com/benleb/ad-automoli
"""

__version__ = "0.8.3"

from copy import deepcopy
from datetime import time
from pprint import pformat
from sys import version_info
from typing import Any, Dict, Iterable, List, Optional, Set, Union

import hassapi as hass


APP_NAME = "AutoMoLi"
APP_ICON = "💡"

ON_ICON = APP_ICON
OFF_ICON = "🌑"
DAYTIME_SWITCH_ICON = "⏰"

# default values
DEFAULT_NAME = "daytime"
DEFAULT_LIGHT_SETTING = 100
DEFAULT_DELAY = 150
DEFAULT_DAYTIMES = [
    dict(starttime="05:30", name="morning", light=25),
    dict(starttime="07:30", name="day", light=100),
    dict(starttime="20:30", name="evening", light=90),
    dict(starttime="22:30", name="night", light=0),
]

EVENT_MOTION_XIAOMI = "xiaomi_aqara.motion"

KEYWORD_LIGHTS = "light."
KEYWORD_MOTION = "binary_sensor.motion_sensor_"
KEYWORD_HUMIDITY = "sensor.humidity_"
KEYWORD_ILLUMINANCE = "sensor.illumination_"

KEYWORDS = {
    "humidity": "sensor.humidity_",
    "illuminance": "sensor.illumination_",
    "light": "light.",
    "motion": "binary_sensor.motion_sensor_",
}

SENSORS_REQUIRED = ["motion"]
SENSORS_OPTIONAL = ["humidity", "illuminance"]

RANDOMIZE_SEC = 5


# version checks
py3_or_higher = version_info.major >= 3
py37_or_higher = py3_or_higher and version_info.minor >= 7
py38_or_higher = py3_or_higher and version_info.minor >= 8


def hl(text: Union[int, float, str]) -> str:
    return f"\033[1m{text}\033[0m"


def hl_entity(entity: str) -> str:
    domain, entity = entity.split(".")
    return f"{domain}.{hl(entity)}"


class AutoMoLi(hass.Hass):  # type: ignore
    """Automatic Motion Lights."""

    def lg(self, msg: str, *args: Any, icon: Optional[str] = None, repeat: int = 1, **kwargs: Any) -> None:
        kwargs.setdefault("ascii_encode", False)
        message = f"{f'{icon} ' if icon else ' '}{msg}"
        _ = [self.log(message, *args, **kwargs) for _ in range(repeat)]

    def listr(self, list_or_string: Union[List[str], Set[str], str], entities_exist: bool = True) -> Set[str]:
        entity_list: List[str] = []

        if isinstance(list_or_string, str):
            entity_list.append(list_or_string)
        elif isinstance(list_or_string, list) or isinstance(list_or_string, set):
            entity_list += list_or_string
        elif list_or_string:
            self.lg(f"{list_or_string} is of type {type(list_or_string)} and not 'Union[List[str], Set[str], str]'")

        return set(filter(self.entity_exists, entity_list) if entities_exist else entity_list)

    async def initialize(self) -> None:
        """Initialize a room with AutoMoLi."""
        self.icon = APP_ICON

        # get a real dict for the configuration
        self.args = dict(self.args)

        # python version check
        if not py38_or_higher:
            icon_alert = "⚠️"
            self.lg("", icon=icon_alert)
            self.lg("")
            self.lg(f" please update to {hl('Python >= 3.8')}! 🤪", icon=icon_alert)
            self.lg("")
            self.lg("", icon=icon_alert)
        if not py37_or_higher:
            raise ValueError

        # set room
        self.room = str(self.args.pop("room"))

        # general delay
        self.delay = int(self.args.pop("delay", DEFAULT_DELAY))

        # state values
        self.states = {
            "motion_on": self.args.pop("motion_state_on", None),
            "motion_off": self.args.pop("motion_state_off", None),
        }

        # threshold values
        self.thresholds = {
            "humidity": self.args.pop("humidity_threshold", None),
            "illuminance": self.args.pop("illuminance_threshold", None),
        }

        # on/off switch via input.boolean
        self.disable_switch_entities: Set[str] = self.listr(self.args.pop("disable_switch_entities", set()))
        self.disable_switch_states: Set[str] = self.listr(self.args.pop("disable_switch_states", set(["off"])))

        self.disable_hue_groups: bool = self.args.pop("disable_hue_groups", False)

        # eol of the old option name
        if "disable_switch_entity" in self.args:
            icon_alert = "⚠️"
            self.lg("", icon=icon_alert)
            self.lg("", icon=icon_alert)
            self.lg("", icon=icon_alert)
            self.lg(
                f" please migrate {hl('disable_switch_entity')} to {hl('disable_switch_entities')}", icon=icon_alert
            )
            self.lg("", icon=icon_alert)
            self.lg("", icon=icon_alert)
            self.lg("", icon=icon_alert)
            self.args.pop("disable_switch_entity")
            return

        # currently active daytime settings
        self.active: Dict[str, Union[int, str]] = {}

        # define light entities switched by automoli
        self.lights: Set[str] = self.args.pop("lights", set())
        if not self.lights:
            room_light_group = f"light.{self.room}"
            if await self.entity_exists(room_light_group):
                self.lights.add(room_light_group)
            else:
                self.lights.update(await self.find_sensors(KEYWORD_LIGHTS))

        # sensors
        self.sensors: Dict[str, Any] = {}

        # enumerate sensors for motion detection
        self.sensors["motion"] = self.listr(self.args.pop("motion", await self.find_sensors(KEYWORD_MOTION)))

        # requirements check
        if not self.lights or not self.sensors["motion"]:
            self.lg("")
            self.lg(
                f"{hl('No lights/sensors')} given and none found with name: "
                f"'{hl(KEYWORD_LIGHTS)}*{hl(self.room)}*' or '{hl(KEYWORD_MOTION)}*{hl(self.room)}*'",
                icon="⚠️ ",
            )
            self.lg("")
            self.lg("  docs: https://github.com/benleb/ad-automoli")
            self.lg("")
            return

        # enumerate optional sensors & disable optional features if sensors are not available
        for sensor_type in SENSORS_OPTIONAL:

            if sensor_type in self.thresholds and self.thresholds[sensor_type]:
                self.sensors[sensor_type] = self.listr(
                    self.args.pop(sensor_type, await self.find_sensors(KEYWORDS[sensor_type]))
                )

            else:
                self.lg(
                    f"No {sensor_type} sensors → disabling features based on {sensor_type}"
                    f" - {self.thresholds[sensor_type]}.",
                    level="DEBUG",
                )
                del self.thresholds[sensor_type]

        # use user-defined daytimes if available
        daytimes = await self.build_daytimes(self.args.pop("daytimes", DEFAULT_DAYTIMES))

        # set up event listener for each sensor
        for sensor in self.sensors["motion"]:

            # listen to xiaomi sensors by default
            if not any([self.states["motion_on"], self.states["motion_off"]]):
                self.lg("no motion states configured - using event listener", level="DEBUG")
                await self.listen_event(self.motion_event, event=EVENT_MOTION_XIAOMI, entity_id=sensor)

            # on/off-only sensors without events on every motion
            elif all([self.states["motion_on"], self.states["motion_off"]]):
                self.lg("both motion states configured - using state listener", level="DEBUG")
                await self.listen_state(self.motion_detected, entity=sensor, new=self.states["motion_on"])
                await self.listen_state(self.motion_cleared, entity=sensor, new=self.states["motion_off"])

        # callback handles to switch lights off
        self.handles: Set[str] = set()

        self.args.update(
            {
                "room": self.room.capitalize(),
                "delay": self.delay,
                "active_daytime": self.active_daytime,
                "daytimes": daytimes,
                "lights": self.lights,
                "sensors": self.sensors,
                "disable_hue_groups": self.disable_hue_groups,
            }
        )

        if self.disable_switch_entities:
            self.args.update({"disable_switch_entities": self.disable_switch_entities})
            self.args.update({"disable_switch_states": self.disable_switch_states})

        # show parsed config
        self.show_info(self.args)

        await self.refresh_timer()

    async def switch_daytime(self, kwargs: Dict[str, Any]) -> None:
        """Set new light settings according to daytime."""
        daytime = kwargs.get("daytime")

        if daytime is not None:
            self.active = daytime
            if not kwargs.get("initial"):

                delay = daytime["delay"]
                light_setting = daytime["light_setting"]
                if isinstance(light_setting, str):
                    is_scene = True
                    # if its a ha scene, remove the "scene." part
                    if "." in light_setting:
                        light_setting = (light_setting.split("."))[1]
                else:
                    is_scene = False

                self.lg(
                    f"set {hl(self.room.capitalize())} to {hl(daytime['daytime'])} → "
                    f"{'scene' if is_scene else 'brightness'}: {hl(light_setting)}"
                    f"{'' if is_scene else '%'}, delay: {hl(delay)}sec",
                    icon=DAYTIME_SWITCH_ICON,
                )

    async def motion_cleared(self, entity: str, attribute: str, old: str, new: str, kwargs: Dict[str, Any]) -> None:
        # starte the timer if motion is cleared
        self.lg(f"motion cleared: {entity} changed {attribute} from {old} to {new}", level="DEBUG")

        if all([await self.get_state(sensor) == self.states["motion_off"] for sensor in self.sensors["motion"]]):
            # all motion sensors off, starting timer
            await self.refresh_timer()
        else:
            # cancel scheduled callbacks
            await self.clear_handles(deepcopy(self.handles))

    async def motion_detected(self, entity: str, attribute: str, old: str, new: str, kwargs: Dict[str, Any]) -> None:
        # wrapper function

        self.lg(f"motion detected: {entity} changed {attribute} from {old} to {new}", level="DEBUG")

        # cancel scheduled callbacks
        await self.clear_handles(deepcopy(self.handles))

        self.lg("handles cleared and cancelled all scheduled timers", level="DEBUG")

        # calling motion event handler
        data: Dict[str, Any] = {"entity_id": entity, "new": new, "old": old}
        await self.motion_event("state_changed_detection", data, kwargs)

    async def motion_event(self, event: str, data: Dict[str, str], kwargs: Dict[str, Any]) -> None:
        """Handle motion events."""
        self.lg(f"received '{event}' event from " f"'{data['entity_id'].replace(KEYWORD_MOTION, '')}'", level="DEBUG")

        # check if automoli is disabled via home assistant entity
        if await self.is_disabled():
            return

        # turn on the lights if not already
        if not any([await self.get_state(light) == "on" for light in self.lights]):
            await self.lights_on()
        else:
            self.lg(f"light in {self.room.capitalize()} already on → refreshing the timer", level="DEBUG")

        if event != "state_changed_detection":
            await self.refresh_timer()

    async def clear_handles(self, handles: Set[str]) -> None:
        """clear scheduled timers/callbacks."""
        self.handles.clear()
        [await self.cancel_timer(handle) for handle in handles]

    async def refresh_timer(self) -> None:
        """refresh delay timer."""

        # cancel scheduled callbacks
        await self.clear_handles(deepcopy(self.handles))

        # if no delay is set or delay = 0, lights will not switched off by AutoMoLi
        if delay := self.active.get("delay"):
            # schedule new callback
            self.handles.add(await self.run_in(self.lights_off, delay))

    async def is_disabled(self) -> bool:
        """check if automoli is disabled via home assistant entity"""
        for entity in self.disable_switch_entities:
            if (state := await self.get_state(entity, copy=False)) and state in self.disable_switch_states:
                self.lg(f"{APP_NAME} is disabled by {entity} with {state = }",)
                return True

        return False

    async def lights_on(self) -> None:
        """Turn on the lights."""
        if illuminance_threshold := self.thresholds.get("illuminance"):

            # the "eco mode" check
            for sensor in self.sensors["illuminance"]:
                try:
                    if (illuminance := float(await self.get_state(sensor))) >= illuminance_threshold:
                        self.lg(
                            f"According to {hl(sensor)} its already bright enough ¯\\_(ツ)_/¯"
                            f" | {illuminance} >= {illuminance_threshold}"
                        )
                        return

                except ValueError as error:
                    self.lg(f"could not parse illuminance '{await self.get_state(sensor)}' from '{sensor}': {error}")
                    return

        if (light_setting := self.active.get("light_setting")) and isinstance(light_setting, str):

            # last check until we switch the lights on... really!
            if any([await self.get_state(light) == "on" for light in self.lights]):
                self.lg("¯\\_(ツ)_/¯")
                return

            for entity in self.lights:

                if self.active["is_hue_group"] and await self.get_state(entity_id=entity, attribute="is_hue_group"):
                    await self.call_service(
                        "hue/hue_activate_scene", group_name=await self.friendly_name(entity), scene_name=light_setting,
                    )
                    continue

                item = light_setting if light_setting.startswith("scene.") else entity

                await self.call_service("homeassistant/turn_on", entity_id=item)

            self.lg(
                f"{hl(self.room.capitalize())} turned {hl(f'on')} → "
                f"{'hue' if self.active['is_hue_group'] else 'ha'} scene: "
                f"{hl(light_setting.replace('scene.', ''))}"
                f" | delay: {hl(self.active['delay'])}sec",
                icon=ON_ICON,
            )

        elif isinstance(self.active["light_setting"], int):

            if self.active["light_setting"] == 0:
                await self.lights_off({})

            else:
                # last check until we switch the lights on... really!
                if any([await self.get_state(light) == "on" for light in self.lights]):
                    self.lg("¯\\_(ツ)_/¯")
                    return

                for entity in self.lights:
                    if entity.startswith("switch."):
                        await self.call_service("homeassistant/turn_on", entity_id=entity)
                    else:
                        await self.call_service(
                            "homeassistant/turn_on", entity_id=entity, brightness_pct=self.active["light_setting"],
                        )

                        self.lg(
                            f"{hl(self.room.capitalize())} turned {hl(f'on')} → "
                            f"brightness: {hl(self.active['light_setting'])}%"
                            f" | delay: {hl(self.active['delay'])}sec",
                            icon=ON_ICON,
                        )

        else:
            raise ValueError(f"invalid brightness/scene: {self.active['light_setting']!s} " f"in {self.room}")

    async def lights_off(self, kwargs: Dict[str, Any]) -> None:
        """Turn off the lights."""

        # check if automoli is disabled via home assistant entity
        if await self.is_disabled():
            return

        # the "shower case" check
        if humidity_threshold := self.thresholds.get("humidity"):
            for sensor in self.sensors["humidity"]:
                try:
                    current_humidity = float(await self.get_state(sensor))
                except ValueError as error:
                    self.lg(f"self.get_state(sensor) raised a ValueError: {error}", level="ERROR")
                    continue

                if current_humidity >= humidity_threshold:
                    # blocker.append(sensor)
                    await self.refresh_timer()
                    self.lg(
                        f"🛁 no motion in {hl(self.room.capitalize())} since "
                        f"{hl(self.active['delay'])}sec → "
                        f"but {hl(current_humidity)}%RH > "
                        f"{hl(humidity_threshold)}%RH"
                    )
                    return

        # cancel scheduled callbacks
        await self.clear_handles(deepcopy(self.handles))

        if any([await self.get_state(entity) == "on" for entity in self.lights]):
            for entity in self.lights:
                await self.call_service("homeassistant/turn_off", entity_id=entity)
            self.lg(
                f"no motion in {hl(self.room.capitalize())} since "
                f"{hl(self.active['delay'])}sec → turned {hl(f'off')}",
                icon=OFF_ICON,
            )

            # experimental | reset for xiaomi "super motion" sensors | idea from @wernerhp
            # app: https://github.com/wernerhp/appdaemon_aqara_motion_sensors
            # mod: https://community.smartthings.com/t/making-xiaomi-motion-sensor-a-super-motion-sensor/139806
            for sensor in self.sensors["motion"]:
                await self.set_state(
                    sensor,
                    state="off",
                    attributes=(await self.get_state(sensor, attribute="all")).get("attributes", {}),
                )

    async def find_sensors(self, keyword: str) -> List[str]:
        """Find sensors by looking for a keyword in the friendly_name."""
        return [
            sensor
            for sensor in await self.get_state()
            if keyword in sensor and self.room in (await self.friendly_name(sensor)).lower().replace("ü", "u")
        ]

    async def build_daytimes(self, daytimes: List[Any]) -> Optional[List[Dict[str, Union[int, str]]]]:
        starttimes: Set[time] = set()

        for idx, daytime in enumerate(daytimes):
            dt_name = daytime.get("name", f"{DEFAULT_NAME}_{idx}")
            dt_delay = daytime.get("delay", self.delay)
            dt_light_setting = daytime.get("light", DEFAULT_LIGHT_SETTING)
            if self.disable_hue_groups:
                dt_is_hue_group = False
            else:
                dt_is_hue_group = (
                    isinstance(dt_light_setting, str)
                    and not dt_light_setting.startswith("scene.")
                    and any(
                        [await self.get_state(entity_id=entity, attribute="is_hue_group") for entity in self.lights]
                    )
                )

            dt_start: time
            try:
                dt_start = await self.parse_time(daytime.get("starttime") + ":00", aware=True)
            except ValueError as error:
                raise ValueError(f"missing start time in daytime '{dt_name}': {error}")

            # configuration for this daytime
            daytime = dict(
                daytime=dt_name,
                delay=dt_delay,
                starttime=dt_start.isoformat(),  # datetime is not serializable
                light_setting=dt_light_setting,
                is_hue_group=dt_is_hue_group,
            )

            # info about next daytime
            next_dt_start = time.fromisoformat(str(daytimes[(idx + 1) % len(daytimes)].get("starttime")))

            # collect all start times for sanity check
            if dt_start in starttimes:
                raise ValueError(f"Start times of all daytimes have to be unique! " f"Duplicate found: {dt_start}",)

            starttimes.add(dt_start)

            # check if this daytime should ne active now
            if await self.now_is_between(str(dt_start), str(next_dt_start)):
                await self.switch_daytime(dict(daytime=daytime, initial=True))
                self.active_daytime = daytime.get("daytime")

            # schedule callbacks for daytime switching
            await self.run_daily(
                self.switch_daytime,
                dt_start,
                random_start=-RANDOMIZE_SEC,
                random_end=RANDOMIZE_SEC,
                **dict(daytime=daytime),
            )

        return daytimes

    def show_info(self, config: Optional[Dict[str, Any]] = None) -> None:
        # check if a room is given

        if config:
            self.config = config

        if not self.config:
            self.lg("no configuration available", icon="‼️", level="ERROR")
            return

        room = ""
        if "room" in self.config:
            room = f" · {hl(self.config['room'].capitalize())}"

        self.lg("")
        self.lg(f"{hl(APP_NAME)}{room}", icon=self.icon)
        self.lg("")

        listeners = self.config.pop("listeners", None)

        for key, value in self.config.items():

            # hide "internal keys" when displaying config
            if key in ["module", "class"] or key.startswith("_"):
                continue

            if isinstance(value, list) or isinstance(value, set):
                self.print_collection(key, value, 2)
            elif isinstance(value, dict):
                self.print_collection(key, value, 2)
            else:
                self._print_cfg_setting(key, value, 2)

        if listeners:
            self.lg("  event listeners:")
            for listener in sorted(listeners):
                self.lg(f"    · {hl(listener)}")

        self.lg("")

    def print_collection(self, key: str, collection: Iterable[Any], indentation: int = 0) -> None:

        self.lg(f"{indentation * ' '}{key}:")
        indentation = indentation + 2

        for item in collection:
            indent = indentation * " "

            if isinstance(item, dict):

                if "name" in item:
                    self.print_collection(item.pop("name", ""), item, indentation)
                else:
                    self.lg(f"{indent}{hl(pformat(item, compact=True))}")

            elif isinstance(collection, dict):

                if isinstance(collection[item], set):
                    self.print_collection(item, collection[item], indentation)
                else:
                    self._print_cfg_setting(item, collection[item], indentation)

            else:
                self.lg(f"{indent}· {hl(item)}")

    def _print_cfg_setting(self, key: str, value: Union[int, str], indentation: int) -> None:
        unit = prefix = ""
        indent = indentation * " "

        # legacy way
        if key == "delay" and isinstance(value, int):
            unit = "min"
            min_value = f"{int(value / 60)}:{int(value % 60):02d}"
            self.lg(f"{indent}{key}: {prefix}{hl(min_value)}{unit} ≈ " f"{hl(value)}sec", ascii_encode=False)

        else:
            if "_units" in self.config and key in self.config["_units"]:
                unit = self.config["_units"][key]
            if "_prefixes" in self.config and key in self.config["_prefixes"]:
                prefix = self.config["_prefixes"][key]

            self.lg(f"{indent}{key}: {prefix}{hl(value)}{unit}")
